include/exadg/incompressible_navier_stokes/preconditioners/multigrid_preconditioner_momentum.cpp

```diff
@@ -142,8 +142,23 @@ MultigridPreconditioner<dim, Number>::fill_matrix_free_data(
 
   matrix_free_data.insert_dof_handler(&(*this->dof_handlers[level]), "std_dof_handler");
   matrix_free_data.insert_constraint(&(*this->constraints[level]), "std_dof_handler");
-  matrix_free_data.insert_quadrature(dealii::QGauss<1>(this->level_info[level].degree() + 1),
-                                     "std_quadrature");
+  if(this->dof_handlers[level]->get_triangulation().all_reference_cells_are_hyper_cube())
+  {
+    matrix_free_data.insert_quadrature(dealii::QGauss<1>(this->level_info[level].degree() + 1),
+                                       "std_quadrature");
+  }
+  else if(this->dof_handlers[level]->get_triangulation().all_reference_cells_are_simplex())
+  {
+    matrix_free_data.insert_quadrature(dealii::QGaussSimplex<dim>(this->level_info[level].degree() +
+                                                                  1),
+                                       "std_quadrature");
+  }
+  else
+  {
+    AssertThrow(false,
+                dealii::ExcMessage("Only pure hypercube or pure simplex meshes are implemented for "
+                                   "MomentumPreconditioner::Multigrid."));
```

Comment:
The notion `MomentumPreconditione::Multigrid` is not known by this class (it is introduced in an outer software layer like `IncNS::Parameters`). Since this class can be used independently of the parameter `MomentumPreconditioner`, I think this name should not appear here in the error message. I would instead write the class name, `IncNS::MultigridPreconditioner`.

Reply:
Makes sense. Just for context, this assert was my inspiration:
https://github.com/exadg/exadg/blob/0f6563a5139715382f117cfa731f03e0d820353e/include/exadg/structure/preconditioners/multigrid_preconditioner.cpp#L141-L144

I am not familiar with that part of the code, thus I don't know if the wording of the assert is correct there. Maybe it needs to be reworded, as well?

Reply:
`Structure::MultigridPreconditioner` in the assert that you mention above fits to my comment above. In the assert, we mention the class that implements the functionality (`Structure::MultigridPreconditioner` or `IncNS::MultigridPreconditioner`), not the parameter in `MyModule::Parameters`.